collector-http/src/main/java/zipkin2/collector/otel/http/OpenTelemetryHttpCollector.java

```diff
@@ -68,31 +93,86 @@ public static final class Builder extends CollectorComponent.Builder {
     metrics = builder.metrics;
   }
 
-  @Override public OpenTelemetryHttpCollector start() {
+  @Override
+  public OpenTelemetryHttpCollector start() {
     return this;
   }
 
-  @Override public String toString() {
+  @Override
+  public String toString() {
     return "OpenTelemetryHttpCollector{}";
   }
 
   /**
    * Reconfigures the service per https://opentelemetry.io/docs/specs/otlp/#otlphttp-request
    */
-  @Override public void reconfigure(ServerBuilder sb) {
+  @Override
+  public void reconfigure(ServerBuilder sb) {
     sb.service("/v1/traces", new HttpService(this));
   }
 
   static final class HttpService extends AbstractHttpService {
+
     final OpenTelemetryHttpCollector collector;
 
     HttpService(OpenTelemetryHttpCollector collector) {
       this.collector = collector;
     }
 
-    @Override protected HttpResponse doPost(ServiceRequestContext ctx, HttpRequest req)
+    @Override
+    protected HttpResponse doPost(ServiceRequestContext ctx, HttpRequest req)
         throws Exception {
-      throw new RuntimeException("Implement me!");
+      CompletableCallback result = new CompletableCallback();
+      req.aggregate(AggregationOptions.usePooledObjects(ByteBufAllocator.DEFAULT, ctx.eventLoop()
+      )).handle((msg, t) -> {
+        if (t != null) {
+          result.onError(t);
+          return null;
+        }
+        try (HttpData content = msg.content()) {
+          if (content.isEmpty()) {
+            result.onSuccess(null);
+            return null;
+          }
+
+          ByteBuf byteBuf = content.byteBuf();
+          final ByteBuffer nioBuffer = byteBuf.nioBuffer();
+          try (ReadBuffer readBuffer = ReadBuffer.wrapUnsafe(nioBuffer)) {
+            try {
+              InputStream inputStream = readBuffer;
+              if ("gzip".equalsIgnoreCase(req.headers().get("content-encoding"))) {
+                inputStream = new GZIPInputStream(content.toInputStream());
+              } else if ("base64".equalsIgnoreCase(req.headers().get("content-encoding"))) {
+                inputStream = Base64.getDecoder().wrap(content.toInputStream());
+              }
+              ExportTraceServiceRequest request = ExportTraceServiceRequest.parseFrom(inputStream);
```

Comment:
Without needing to decode input streams, you should be able to parse the `ByteBuffer` directly with `CodedInputStream.newInstance(nioBuffer)`

Reply:
How would that look like? I don't want to misuse the API :grimacing:

Reply:
Assuming I didn't miss anything, I think it's just that

```java
ExportTraceServiceRequest request = ExportTraceServiceRequest.parseFrom(CodedInputStream.newInstance(nioBuffer));
```

Don't need `ReadBuffer`, it's a `InputStream` so currently the parsing is going through protobuf's slow `InputStream` parsing path

Reply:
Ah yeah that's the least lines but I think `CodedInputStream.newInstance(content.byteBuf().nioBuffer())` isn't too bad and avoids going through the `InputStream` version of protobuf parsing. It's not terrible but since we've already aggregated, better to avoid it.

Reply:
OK, fixed that. Thank you